Fixed #30196 -- Made FileResponse always set Content-Disposition header.

ShingenPizza opened the pull request:
FileResponse currently sets the Content-Disposition header only if as_attachment is true.

Setting it explicitly to, e.g. 'inline; filename="example.png"' in the other case would allow the browser to set a default name for that inline file in case a user attempts to download it with the 'Save image as...' option.

That filename value is also visible in the title of the tab when image is being viewed directly in Firefox (at least v56).

Created a ticket here: https://code.djangoproject.com/ticket/30196

charettes commented on django/http/response.py:
```diff
@@ -436,15 +436,15 @@ def set_headers(self, filelike):
+                filename.encode('ascii')
+                file_expr = f'filename="{filename}"'
+            except UnicodeEncodeError:
+                file_expr = f"filename*=utf-8''{quote(filename)}"
```
Stick to `.format` for now. Until we settle to switch to f strings for the whole code base we should stick one format.

ShingenPizza replied:
Alright, done.
Didn't think it'll be an issue given that only ~10 lines below there's still %-formatting used... :)

charettes replied:
yeah `.format` crept in while we were supposed to stick to `%`. anyway lets make sure we don't have three ways of formatting strings across the code base 😅 .

charettes commented:
As mentioned on Trac tests should be added as well.

ShingenPizza commented:
(I'll just add this here too for a possibly quicker response)
Sorry for _beginner_ questions, but: as in, you want me to add some unit tests?
No problem, but wouldn't a single test be enough?
All this change does, after all, is setting Content-Disposition when as_attachment=False, so checking if it works shouldn't require more.
I also noticed there's no unit test checking if setting a custom filename works, so i added it to the same one test.
What else to add?

timgraham commented:
Use your judgement and add a single test if you feel it covers the change. Additional test coverage unrelated to the fix is okay but should be added in a separate commit or PR.

ShingenPizza commented:
Alright.
If so, then i just made it similar to the as_attachment=True unit test, and packed the changes into a single commit.

Once this is finished i will create another PR for an additional test of custom filenames - do i need to create a ticket on Trac for that too?

timgraham commented:
Additional test coverage doesn't require a ticket.

timgraham commented:
It seems that documentation changes are also required. See a177f854c34718e473bcd0a2dc6c4fd935c8e327.

ShingenPizza commented:
Alright, i expanded the descriptions of as_attachment in docs/howto/outputting-pdf.txt and docs/ref/request-response.txt. Will that suffice?

I added setting Content-Disposition even when filename isn't known (to just 'inline' or 'attachment').
I added the line of difference between my unit test and one other that i failed to notice before, to that old one.
I added an additional check to unnamed binary content unit test (test_file_from_buffer_response) to see whether it correctly sets Content-Disposition to just 'inline'.

carltongibson commented on tests/responses/test_fileresponse.py:
```diff
@@ -14,12 +14,14 @@ def test_file_from_disk_response(self):
         response = FileResponse(io.BytesIO(b'binary content'))
         self.assertEqual(response['Content-Length'], '14')
         self.assertEqual(response['Content-Type'], 'application/octet-stream')
+        self.assertEqual(response['Content-Disposition'], 'inline')
```
Is this the behaviour we want when there's no filename? 

[From the RFC](https://tools.ietf.org/html/rfc6266#section-4.2):

> ...if it matches "inline" (case-insensitively), this
   implies default processing.  Therefore, the disposition type "inline"
   is only useful when it is augmented with additional parameters, such
   as the filename...

i.e. if we're not augmenting it, it's not useful. 

[The examples on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#Syntax) don't show the `inline; filename` syntax as an option. Happy that it is valid, but do we want to **always** include it?

ShingenPizza replied:
I agree that it's unnecessary.
I think I did it this way to make the code nice and clean, because the alternative I was thinking about wouldn't be as clean, but now I realized there's a clean enough way to do it without setting it in each case, so I changed it appropriately.

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -1091,8 +1091,10 @@ Attributes
+    If ``as_attachment=True``, the ``Content-Disposition`` header is set to
+    ``attachment``, which asks the browser to offer the file to the user as a
+    download. If it's ``False``, ``Content-Disposition`` is set to ``inline``,
+    which tells the browser to handle it in the way that it's configured to.
```
This seems a bit clunky. Maybe something like: 

> Otherwise a default `inline` `Content-Disposition` header will only be set if a `filename` is available. 

(This assuming we **don't** always set it, as per other comment.) 

**Either way**, having discussed `Content-Disposition` here, do we not need to adjust the `...headers are automatically set...` sentence below to remove it from there?

ShingenPizza replied:
Changed to suggested sentence.

> Either way, having discussed `Content-Disposition` here, do we not need to adjust the `...headers are automatically set...` sentence below to remove it from there?

Yeah, I agree. Removed it.

carltongibson commented on docs/howto/outputting-pdf.txt:
```diff
@@ -78,11 +78,12 @@ mention:
+  appropriate ``Content-Disposition`` header to ``attachment``, and that tells
+  Web browsers to pop-up a dialog box prompting/confirming how to handle the
+  document even if a default is set on the machine. If the ``as_attachment``
+  parameter is ommited or ``False``, ``Content-Disposition`` is set to
```
Not sure about the _"ommited or `False`"_. I kind of think just _"Otherwise browsers will handle..."_ would read just as well. 

(Given that, I'm not sure we need this change here...?)

ShingenPizza replied:
Suits me, reverted.

ShingenPizza commented:
Speaking of further changes to the docs and such...

    .. method:: FileResponse.set_headers(open_file)
        This method is automatically called during the response initialization and
        set various headers (``Content-Length``, ``Content-Type``, and
        ``Content-Disposition``) depending on ``open_file``.

1. Technically, `FileResponse.set_headers()` does not have an `open_file` parameter, but a `filelike` one. `open_file` is described above in docs...
>
    .. class:: FileResponse(open_file, as_attachment=False, filename='', **kwargs)
        (...)
        If ``open_file`` doesn't have a name or if the name of ``open_file`` isn't
        appropriate, provide a custom file name using the ``filename``  parameter.
... but that also doesn't appear in the code. `FileResponse.__init__()` uses `args*` which eventually end as `streaming_content` of `StreamingHttpResponse`. Is that all on purpose, or should it be altered?

2. What even is the purpose of the `FileResponse.set_headers()`?
- Being a regular method its allows to be accidentally executed and thus to change these headers to match a different file than the one that is being returned, which is rather not a good idea. (Actually, `Content-Type` can't change once it's set to anything but a string starting from `'text/html'`).
- Additionally, I just noticed that in some cases it breaks, for example if you use it like this: `FileResponse(io.BytesIO(b'binary content'), filename=__file__)` (more generally: `FileResponse(`_filelike object without `name` attribute_`, filename=`_any string recognized by os.path as absolute_`)`).

Thus, I suggest:
- move the contents of `.set_headers()` to `._set_streaming_content()`, or make it _private_ (`._set_headers()`).
- ignore `self.filename` at setting `Content-Length` by removing `filename = filename if (isinstance(filename, str) and filename) else self.filename` - after all, it's a filename for `Content-Disposition`, and if I want to return one file under a name of another, it should not affect other headers.
- remove `if self.get('Content-Type', '').startswith('text/html'):` because, current value shouldn't be relevant if you want to set a new one, and also it may be bug-prone if you somehow manage to set `Content-Type` to anything else before .

I am aware that even if it's not a bad idea to change it the way I suggest or similarly, it would be done in a separate PR, but I'm asking about it upfront.

carltongibson commented on django/http/response.py:
```diff
@@ -436,15 +436,17 @@ def set_headers(self, filelike):
+                file_expr = "filename*=utf-8''{}".format(quote(filename))
+            self['Content-Disposition'] = '{}; {}'.format(disposition, file_expr)
+        elif self.as_attachment:
+            self['Content-Disposition'] = 'attachment'
```
Can we add a test case for this branch please?

ShingenPizza replied:
Done.

carltongibson commented:
@ShingenPizza re https://github.com/django/django/pull/11011#issuecomment-475827865...

**On 1**: the signatures don't have to match exactly. `open_file` is the usual case. `filelike`... well... OK, but maybe not as clear to some. For the `*args` case I think strictly less clear (what else are you going to pass?) **But** this is just my initial take: if you want to propose a concrete change, very happy to look at that. 

**On 2**: Giving a method a `_` doesn't stop anyone calling it. You'd have had to go looking and wanting to call this, so I'm not convinced about the rename... The behavioral changes are possibles, yes... If you want to put forward test cases showing incorrect behavior then we'd always look at those.

ShingenPizza commented:
> **On 1**: the signatures don't have to match exactly. `open_file` is the usual case. `filelike`... well... OK, but maybe not as clear to some. For the `*args` case I think strictly less clear (what else are you going to pass?) **But** this is just my initial take: if you want to propose a concrete change, very happy to look at that.

It was more a way to check if it's intentional. If you guys don't mind it being the way it is, that's fine with me.

> **On 2**: Giving a method a `_` doesn't stop anyone calling it. You'd have had to go looking and wanting to call this, so I'm not convinced about the rename...

It doesn't stop anyone, absolutely. However, it symbolically marks it to not be explicitly called from outside of the class - and technically we could use `__` (double `_`) if making it truly (well, almost) private was our concern :P
I disagree that you _have to go looking_, however, since... it's mentioned right there in the docs.
I remember seeing it there for the first time and being a little bit confused, because on one hand it states that it sets the headers (my immediate thought being _so I have to call it, right?_), but on the other it says it's called automatically... and after a further investigation it, i'd say, **should not** be called more than once, so never from the outside.
It's fine as a way to easily alter/fix its functionality in a child class (that's what I'm doing now in my little project until my changes here get released), so maybe my idea to move its functionality to `_set_streaming_content` wasn't the best, but I still think it would be good to make it look less like a method _to be called by someone using `FileResponse`_. But of course, it's not my decision to make.

> The behavioral changes are possibles, yes... If you want to put forward test cases showing incorrect behavior then we'd always look at those.

Alright, I'll do that in a separate ticket/PR after we're done here.

rixx commented:
Looks like a pretty polished and useful change to me by now – thank you, @ShingenPizza!

The documentation changes look alright in general, but they don't mention that an appropriate filename is sent in the ``Content-Disposition`` header – this will be clear to everybody who knows the header format off the top of their head, but it might be nice to mention it for everybody else?

ShingenPizza commented:
@rixx 
I guess it's not a bad idea, but I'm unsure how to go about it.
Should we go minimalistic, like the below?
> The filename is appended to this header's value after a semicolon.

Should we mention that it's in a format of `filename="[actual filename]"`?
Should we mention the handling of unicode names?

Not mentioning every case isn't _thorough_, but on the other hand doing so seems like adding a lot of text that isn't really necessary for a regular user - a link to a description of these headers would be easier.

Actually, having said that, is describing the actual values set to the header that we're doing now even necessary?
I'd say that documentation should describe more of what's the _functional_ result of a given parameter and not that much of what exactly happens in the code (what values are set where) - if someone wants to know the latter they can just check out the code.

So I'd be for simplifying the docs, but I'd rather do it with in the other planned PR, and just get this merged already (1.5 months for such a relatively small change is long enough ^^').

rixx commented:
I definitely don't want to hold up this PR, as far as I can see, it's ready for being merged.

As far as the docs go, they currently say:

> Otherwise, a ``Content-Disposition`` header with a value of
>    ``inline`` (the browser default) will be set only if a filename is
>    available.

I liked that you mentioned documenting the results, not the mechanics. How about

> Otherwise, if a filename is available, it will be provided in the response via the ``Content-Disposition`` header.

Or something like that?

ShingenPizza commented:
No, that wouldn't be appropriate, because the 'otherwise' refers to a situation where `as_attachment` is `False`, and the filename is returned also when it's `True` - it's actually even more important then.

I would change the entire section from
>If ``as_attachment=True``, the ``Content-Disposition`` header is set to ``attachment``, which asks the browser to offer the file to the user as a download. Otherwise, a ``Content-Disposition`` header with a value of ``inline`` (the browser default) will be set only if a filename is available.

to simply (the part in the square brackets may not be needed)
>If ``as_attachment=True``, the response will ask the browser to offer the file to the user as a download. [Otherwise, the browser will process the file in the default way.]

and remove
>The ``Content-Length`` and ``Content-Type`` headers are automatically set when they can be guessed from contents of ``open_file``.

or at least rephrase it, because:
1. There's nothing to _guess_ to set `Content-Length`.
2. It doesn't mention you can overwrite the `Content-Type` by `HttpResponseBase`'s `content_type` parameter.

ShingenPizza commented:
Is there anything left that I'm supposed to do to get it pulled? :T

rixx commented:
I'm afraid you'll have to be patient – people are working on PR review, but as you can see from the number of open PRs, that's a big job. Your PR and your ticket have both all the right flags set, so you're bound to receive further review eventually.

ShingenPizza commented:
Alright then, I just wanted to make sure that they're not waiting for me somehow.

And thanks for altering the that ticket a couple of days ago.

carltongibson commented:
This is on the list. (IIRC I don’t think it’s far off...) 

We’re just kind of focused on v2.2.1 this week. 

(Any reviewer can mark it RFC on Trac if you think it’s ready. 🙂)
